src/hyperlane-adapter/HyperlaneSenderAdapter.sol

```diff
@@ -0,0 +1,293 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity 0.8.16;
+
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { IMailbox } from "./interfaces/IMailbox.sol";
+import { IInterchainGasPaymaster } from "./interfaces/IInterchainGasPaymaster.sol";
+import { TypeCasts } from "./libraries/TypeCasts.sol";
+import { Errors } from "./libraries/Errors.sol";
+import { IMessageDispatcher, ISingleMessageDispatcher } from "./interfaces/ISingleMessageDispatcher.sol";
+import { IBatchedMessageDispatcher } from "./interfaces/IBatchedMessageDispatcher.sol";
+import { IMessageExecutor } from "./interfaces/IMessageExecutor.sol";
+import "./libraries/MessageLib.sol";
+
+contract HyperlaneSenderAdapter is ISingleMessageDispatcher, IBatchedMessageDispatcher, Ownable {
+  /// @notice `Mailbox` contract reference.
+  IMailbox public immutable mailbox;
+
+  /// @notice `IGP` contract reference.
+  IInterchainGasPaymaster public igp;
+
+  uint256 public nonce;
+
+  /**
+   * @notice Receiver adapter address for each destination chain.
+   * @dev dstChainId => receiverAdapter address.
+   */
+  mapping(uint256 => IMessageExecutor) public receiverAdapters;
+
+  mapping(uint256 => bool) public isValidChainId;
+
+  /**
+   * @notice Domain identifier for each destination chain.
+   * @dev dstChainId => dstDomainId.
+   */
+  mapping(uint256 => uint32) public destinationDomains;
+
+  /**
+   * @notice Emitted when the IGP is set.
+   * @param paymaster The new IGP for this adapter.
+   */
+  event IgpSet(address indexed paymaster);
+
+  /**
+   * @notice Emitted when a receiver adapter for a destination chain is updated.
+   * @param dstChainId Destination chain identifier.
+   * @param receiverAdapter Address of the receiver adapter.
+   */
+  event ReceiverAdapterUpdated(uint256 dstChainId, IMessageExecutor receiverAdapter);
+
+  /**
+   * @notice Emitted when a domain identifier for a destination chain is updated.
+   * @param dstChainId Destination chain identifier.
+   * @param dstDomainId Destination domain identifier.
+   */
+  event DestinationDomainUpdated(uint256 dstChainId, uint32 dstDomainId);
+
+  /**
+   * @notice HyperlaneSenderAdapter constructor.
+   * @param _mailbox Address of the Hyperlane `Mailbox` contract.
+   */
+  constructor(address _mailbox, address _igp) {
+    if (_mailbox == address(0)) {
+      revert Errors.InvalidMailboxZeroAddress();
+    }
+    mailbox = IMailbox(_mailbox);
+    _setIgp(_igp);
+  }
+
+  /// @dev we narrow mutability (from view to pure) to remove compiler warnings.
+  /// @dev unused parameters are added as comments for legibility.
+  function getMessageFee(
+    uint256 toChainId,
+    address,
+    /* to*/
+    bytes calldata /* data*/
+  ) external view returns (uint256) {
+    uint32 dstDomainId = _getDestinationDomain(toChainId);
+    // destination gasAmount is hardcoded to 500k similar to Wormhole implementation
```

Comment:
Why not making this a configuration option in the constructor (and still default to 500k)

Reply:
Done